build/Xamarin.Build/MSBuild/ILRepack.cs

```diff
@@ -0,0 +1,187 @@
+using System.Text.RegularExpressions;
+
+using ILRepacking;
+
+using Mono.Cecil;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.MSBuild
+{
+    public sealed class ILRepack : Task, ILRepacking.ILogger
+    {
+        [Required]
+        public string [] InputAssemblies { get; set; } = Array.Empty<string> ();
+
+        public string [] SearchDirectories { get; set; } = Array.Empty<string> ();
+
+        public string [] References { get; set; } = Array.Empty<string> ();
+
+        public string OutputAssembly { get; set; }
+
+        public ITaskItem [] NamespaceRenames { get; set; } = Array.Empty<ITaskItem> ();
+
+        public override bool Execute ()
+        {
+            var basePath = Path.GetDirectoryName (BuildEngine.ProjectFileOfTaskNode);
+
+            IEnumerable<string> ExpandPath (IEnumerable<string> paths)
+                => paths.Select (p => Path.GetFullPath (Path.Combine (basePath, p)));
+
+            var searchPaths = new List<string> (ExpandPath (SearchDirectories));
+            var inputAssemblies = new List<string> (ExpandPath (InputAssemblies));
+            var outputAssembly = OutputAssembly ?? InputAssemblies [0];
+
+            foreach (var reference in ExpandPath (References)) {
+                var searchPath = Path.GetDirectoryName (reference);
+
+                if (!IsNetstandardLibrary (reference) &&
+                    !inputAssemblies.Contains (reference, StringComparer.OrdinalIgnoreCase))
+                    inputAssemblies.Add (reference);
+                else if (!searchPaths.Contains (searchPath, StringComparer.OrdinalIgnoreCase))
+                    searchPaths.Add (searchPath);
+            }
+
+            Log.LogMessage (MessageImportance.Normal, "Output Assembly: {0}", outputAssembly);
+            Log.LogMessage (MessageImportance.Normal, "Internalize: {0}", outputAssembly);
+
+            foreach (var searchPath in searchPaths)
+                Log.LogMessage (MessageImportance.Normal, "Search Path: {0}", searchPath);
+
+            foreach (var inputAssembly in inputAssemblies)
+                Log.LogMessage (MessageImportance.Normal, "Input Assembly: {0}", inputAssembly);
+
+            var resolver = new DefaultAssemblyResolver ();
+            foreach (var searchPath in searchPaths)
+                resolver.AddSearchDirectory (searchPath);
+
+            var readerParameters = new ReaderParameters {
+                AssemblyResolver = resolver,
+                ReadingMode = ReadingMode.Immediate
+            };
+
+            var renames = new List<(Regex regex, string replacement)> ();
+            foreach (var rename in NamespaceRenames)
+                renames.Add ((new Regex (rename.ItemSpec), rename.GetMetadata ("Replacement")));
+
+            var typesToRenamespace = new HashSet<string> ();
+            var renamespaceMap = new Dictionary<string, string> ();
+
+            for (int i = 1; i < inputAssemblies.Count; i++) {
```

Comment:
Just checking: we skip the first input assembly because it's the "main" one? X.I.dll for example? And therefore it doesn't need any types renamespaced?

Reply:
Yes. This convention brought to you by ILRepack.